pkg/controllers/state/statenode.go

```diff
@@ -291,6 +290,13 @@ func (in *StateNode) Taints() []v1.Taint {
 			return false
 		})
 	}
+	// Add an in-memory NoSchedule taint for all eventual disruption candidates. This ensures that we don't consider
+	// these nodes as candidates for pending pods being rescheduled from other disruption candidates.
+	if in.EventualDisruptionCandidate() && !lo.ContainsBy(taints, func(t v1.Taint) bool {
+		return t.MatchTaint(&v1beta1.DisruptionCandidateNoScheduleTaint)
+	}) {
+		taints = append(taints, v1beta1.DisruptionCandidateNoScheduleTaint)
```

Comment:
Should we be doing an append or a mutation of the existing taint?

Reply:
This could be added before the tainting controller adds the `PreferNoSchedule` taint, or before cluster state is aware of the change. This eliminates a case where we may run a scheduling simulation and nominate a pod for a drifted node before we become aware of the taint. I don't think this is strictly necessary, but is an improvement.

Reply:
The way this is orchestrated means that someone could theoretically tolerate the PreferNoSchedule taint with their pods, but their pods wouldn't tolerate the NoSchedule, right? We might want to also allow the pods that tolerate PreferNoSchedule to also tolerate this in memory by default.

Reply:
That's a good point, it's currently in a weird state where if you tolerate the taint we apply you still get the weigh-away behavior, but not if you tolerate all effects. I'm wondering if a toleration is the right mechanism to opt out of the weigh away though (vs a separate field we could add down the line). FWIW I think it probably is since, by tolerating the `PreferNoSchedule` taint, there's nothing stopping a pod from scheduling to a drifted node even if we spin up new capacity.